### PR TITLE
Add a way to specify certificate validity period on Darwin.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
This PR includes the change in https://github.com/project-chip/connectedhomeip/pull/26402 to pass the newly added tests, but I'd like to merge it after the other PR merges, for clarity.

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRCertificates.h [hunk 1]
```diff
@@ -44,14 +44,27 @@ NS_ASSUME_NONNULL_BEGIN
  * If fabricID is not nil, it will be included in the subject DN of the
  * certificate.  In this case it must be a valid Matter fabric id.
  *
+ * validityPeriod specifies when the certificate will be valid.
```
Comment (ksperling-apple) [inline src/darwin/Framework/CHIP/MTRCertificates.h:47]:
```suggestion
 * validityPeriod specifies when the certificate will be valid. Note that
 * there is currently no mechanism available in Matter to update or rotate
 * the root certificate of a fabric installed on a device.
```

File: src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.h [hunk 2]
```diff
@@ -102,21 +102,26 @@ class MTROperationalCredentialsDelegate : public chip::Controller::OperationalCr
...
     // Generate an operational DER-encoded X.509 certificate for the given
     // signing certificate and operational public key, using the given fabric
     // id, node id, and CATs.
     static CHIP_ERROR GenerateOperationalCertificate(id<MTRKeypair> signingKeypair, NSData * signingCertificate,
         SecKeyRef operationalPublicKey, NSNumber * fabricId, NSNumber * nodeId, NSSet<NSNumber *> * _Nullable caseAuthenticatedTags,
-        NSData * _Nullable __autoreleasing * _Nonnull operationalCert);
+        NSDateInterval * validityPeriod, NSData * _Nullable __autoreleasing * _Nonnull operationalCert);
+
+    // 10 years.
+    static const uint32_t kCertificateValiditySecs = 10 * 365 * 24 * 60 * 60;
```
Comment (ksperling-apple) [inline src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.h:115]:
```suggestion
    static const uint32_t kCertificateDefaultValiditySecs = 10 * 365 * 24 * 60 * 60;
```

File: src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm [hunk 3]
```diff
@@ -310,12 +311,28 @@
...
+    return ToChipEpochTime(date, epoch);
+}
+
+bool MTROperationalCredentialsDelegate::ToChipEpochTime(NSDate * date, uint32_t & epoch)
 {
-    NSDate * date = [NSDate dateWithTimeIntervalSinceNow:offset];
     NSCalendar * calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
     NSDateComponents * components = [calendar componentsInTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0] fromDate:date];
 
+    if (!CanCastTo<uint16_t>(components.year)) {
```
Comment (ksperling-apple) [inline src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm:329]:
```suggestion
    if (components.year > kChipEpochMaxYear) {
```
kChipEpochMaxYear (2135) is quite a bit smaller than uint16 range

File: src/darwin/Framework/CHIPTests/MTRCertificateTests.m [hunk 4]
```diff
@@ -74,6 +140,88 @@ - (void)testGenerateIntermediateCert
...
+    __auto_type * rootCert = [MTRCertificates createRootCertificate:rootKeys issuerID:nil fabricID:nil error:nil];
+    XCTAssertNotNil(rootCert);
+
+    __auto_type * intermediateKeys = [[MTRTestKeys alloc] init];
+    XCTAssertNotNil(intermediateKeys);
+
+    __auto_type * startDate = [NSDate dateWithTimeIntervalSinceNow:300];
+    // Round down to the nearest second, since the certificate bits will do that
+    // when it computes validity dates.
+    NSTimeInterval seconds = floor([startDate timeIntervalSinceReferenceDate]);
```
Comment (ksperling-apple) [inline src/darwin/Framework/CHIPTests/MTRCertificateTests.m:198]: Pull the rounding logic out into a helper function since it's repeated 3 times?

**Review by ksperling-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm (hunk 3 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm:329]: > kChipEpochMaxYear (2135) is quite a bit smaller than uint16 range

It is, but that validation will be done by CalendarToChipEpochTime.  I'd rather not hardcode that dependency here, and just do the one check _this_ code relies on, which is that the year can be losslessly represented in `uint16_t`.

File: src/darwin/Framework/CHIP/MTRCertificates.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRCertificates.h:47]: Yep, will do.  Also added comments about distantFuture meaning no expiration.

File: src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.h (hunk 2 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.h]: Done.

File: src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm (hunk 3 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm:329]: I don't think there would be any problems coming out of depending on `kChipEpochMaxYear` here, and your error log text is much more helpful than the silent failure of `CalendarToChipEpochTime`, but I'm fine either way.

Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm:329]: Ah, the logging I am happy to fix.

**Review by jmeg-sfy — APPROVED** [review]

**Pull request merged**